src/main/java/io/wgj/ctci/ArrayList.java

```diff
@@ -0,0 +1,105 @@
+package io.wgj.ctci;
+
+import java.util.Iterator;
+
+public class ArrayList <E> implements Iterable<E> {
+  transient Object[] elements;
+  static final int STARTING_CAPACITY = 1;
+  private int size;
+  private int capacity;
+
+  public ArrayList() {
+    elements = new Object[STARTING_CAPACITY];
+    size = 0;
+    capacity = STARTING_CAPACITY;
+  }
+
+  public int size() {
+    return this.size;
+  }
+
+  public boolean add(E e) {
+    if (size >= capacity) {
```

Comment:
I think it's no longer relevant. When I wrote written, I must've thought `size` might be incremented elsewhere, and I was worried I'd confuze number of elements and number of indicies preventing `grow()` from happening. I reasoned that a slightly premature `grow()` was better than a failure to call `grow()`.

Reply:
What I wanted to get at is that the way this class is written it is impossible for ```size``` > ```capacity```, so in some sense you're adding unreachable/untestable code by using ```>=``` instead of just ```==```.

That being said, I see a lot of people still do this in this case because they feel its programming defensively, I guess against cosmic rays flipping RAM bits or something.  Personally when I see something like this I assume I misunderstand the class and go re-read everything, wasting time.